### ISSUE TITLE
[member-ordering] Feature Request: Case-insensitivity

### ISSUE THREAD

**Issue by Zamiell** [typescript-eslint/typescript-eslint#3380]
[search tags: case sensitive insensitive sensitivity insensitivity]

This is a feature request for more functionality in the [`member-ordering`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/member-ordering.md) rule provided by `@typescript-eslint/eslint-plugin`.

(I did a quick search for this and there does not seem to be an issue about it already.)

Currently, the [documentation](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/member-ordering.md) states that it allows users to sort "alphabetically", but that's it. This alphabetical sort is case-sensitive, which means that the following code will throw an error:

```typescript
declare class Foo {
  Bar1(): void;
  bar2(): void;
  Bar3(): void; // Error: Member "Bar3" should be declared before member "bar2".
}
```

This is annoying and makes the alphabetical sort unsuitable for some use-cases.
Would it be possible to add a new option, "alphabetically-case-insensitive" to exist alongside "alphabetically"?

As a separate request, would it also be possible for the user to somehow specify their own sorting function in the ".eslintrc.js" file?

(Use case: I am writing TypeScript definitions that match [this documentation](https://moddingofisaac.com/docs/rep/Vector/), and I want the underscore methods to appear at the top of my definitions like they do in in the docs. However, the current alphabetical sort wants to put the underscore methods at the bottom, so ideally I would want to pass some sorting function that says that "underscore is less than a" or whatnot.)

In closing, thank you for the free linter.

**Comment by bradzacher** [comment]
> As a separate request, would it also be possible for the user to somehow specify their own sorting function in the ".eslintrc.js" file?

No. ESLint config only allows JSON values as the config passes through the JSON schema parser.

-----

Based on your use-case, you don't actually want case insensitive sort, you just want a sort where underscores come first.

Unless there's another usecase you are thinking of when you say "This is annoying and makes the alphabetical sort unsuitable for some use-cases"

**Comment by Zamiell** [comment]
> Based on your use-case, you don't actually want case insensitive sort, you just want a sort where underscores come first.

I think I do want a case-insensitive sort.
For example, I want my TypeScript definitions to match the documentation here:

**Comment by Zamiell** [comment]
Can I submit a PR to add this functionality?

**Comment by bradzacher** [comment]
it's here open for anyone who is so motivated!

**Comment by nbouvrette** [comment]
It would probably make sense to close this issue and re-open a new one regarding the "custom sorting" given that https://github.com/typescript-eslint/typescript-eslint/commit/e3533d5a6293a358b5eb0a6ed17da961a09b0ed3 implemented the case-insensitive part of this multi-ask issue.
